source/utils/algorithms.cpp

```diff
@@ -7,36 +7,52 @@
 
 namespace sim {
 
-// Builds routing table using BFS algorithm starting from start_device,
-// for each device stores the first link that leads to it from the start
+// returns start device routing table
+// Unlike standard BFS that processes nodes one by one, this processes all nodes at the current distance level together.
+// So each iteration is a processing of all all devices at a certain distance (wavefront)
 RoutingTable bfs(std::shared_ptr<IRoutingDevice>& start_device) {
     RoutingTable routing_table;
+
     std::queue<std::shared_ptr<IRoutingDevice>> queue;
+    // All devices with known distance
     std::set<std::shared_ptr<IRoutingDevice>> used;
-    queue.push(start_device);
-
-    while (!queue.empty()) {
-        std::shared_ptr<IRoutingDevice> device = queue.front();
-        queue.pop();
-        if (used.contains(device)) {
-            continue;
+    // Current wavefront
+    std::set<std::shared_ptr<IRoutingDevice>> wave_front;
+    wave_front.insert(start_device);
+    
+    while (!wave_front.empty()) {
+        for (auto device: wave_front) {
+            queue.push(device);
+            used.insert(device);
         }
-        used.insert(device);
-        auto outlinks = device->get_outlinks();
-        for (std::shared_ptr<ILink> link : outlinks) {
-            auto next_hop = link->get_to();
-            auto curr_device = link->get_from();
-            if (used.contains(next_hop)) {
-                continue;
-            }
-            if (curr_device == start_device) {
-                routing_table[next_hop] = link;
-            } else if (!routing_table.contains(next_hop)) {
-                routing_table[next_hop] = routing_table[curr_device];
+        wave_front.clear();
+
+        while (!queue.empty()) {
+            std::shared_ptr<IRoutingDevice> device = queue.front();
+            queue.pop();
+
+            std::set<std::shared_ptr<ILink>> outlinks = device->get_outlinks();
+            for (std::shared_ptr<ILink> link : outlinks) {
+                auto next_hop = link->get_to();
+                auto curr_device = link->get_from();
+                if (used.contains(next_hop)) {
+                    continue;
+                }
+                wave_front.insert(next_hop);
+
+                if (curr_device == start_device) {
+                    routing_table[next_hop][link] = 1;
+                } else {
+                    // Here we get all ways to get to the previous device and add them to next hop
+                    // This part might be called several times for same device
+                    for (auto link_and_paths: routing_table[curr_device]) {
+                        routing_table[next_hop][link_and_paths.first] += link_and_paths.second;
+                    }
```

Comment:
Visual simplification:
```suggestion
                    for (auto [link, paths_count]: routing_table[curr_device]) {
                        routing_table[next_hop][link] += paths_count;
                    }
```